crypto/mlkem/mlkem768.c

```diff
@@ -0,0 +1,1184 @@
+    }
+}
+
+static void vector_add(vector *lhs, const vector *rhs)
+{
+    int i;
+
+    for (i = 0; i < RANK768; i++)
+        scalar_add(&lhs->v[i], &rhs->v[i]);
+}
+
+static void matrix_mult(vector *out, const matrix *m,
+                        const vector *a)
+{
+    int i, j;
+
+    vector_zero(out);
+    for (i = 0; i < RANK768; i++) {
+        for (j = 0; j < RANK768; j++) {
+            scalar product;
+
+            scalar_mult(&product, &m->v[i][j], &a->v[j]);
+            scalar_add(&out->v[i], &product);
+        }
+    }
+}
+
+static void matrix_mult_transpose(vector *out, const matrix *m,
+                                  const vector *a)
+{
+    int i, j;
+
+    vector_zero(out);
+    for (i = 0; i < RANK768; i++) {
+        for (j = 0; j < RANK768; j++) {
+            scalar product;
+
+            scalar_mult(&product, &m->v[j][i], &a->v[j]);
+            scalar_add(&out->v[i], &product);
+        }
+    }
+}
+
+static void scalar_inner_product(scalar *out, const vector *lhs,
+                                 const vector *rhs)
+{
+    int i;
+
+    scalar_zero(out);
+    for (i = 0; i < RANK768; i++) {
+        scalar product;
+
+        scalar_mult(&product, &lhs->v[i], &rhs->v[i]);
+        scalar_add(out, &product);
+    }
+}
+
+/*
+ * Algorithm 6 from the spec. Rejection samples a Keccak stream to get
+ * uniformly distributed elements. This is used for matrix expansion and only
+ * operates on public inputs.
+ */
+static int scalar_from_keccak_vartime(scalar *out, EVP_MD_CTX *mdctx)
+{
+    int done = 0;
+    uint8_t block[168];
+    size_t i;
+    uint16_t d1, d2;
+
+    while (done < DEGREE) {
+        if (!EVP_DigestSqueeze(mdctx, block, sizeof(block)))
+            return 0;
+        for (i = 0; i < sizeof(block) && done < DEGREE; i += 3) {
+            d1 = block[i] + 256 * (block[i + 1] % 16);
+            d2 = block[i + 1] / 16 + 16 * block[i + 2];
+            if (d1 < kPrime)
+                out->c[done++] = d1;
+            if (d2 < kPrime && done < DEGREE)
+                out->c[done++] = d2;
+        }
+    }
+    return 1;
+}
+
+/*
+ * Algorithm 7 from the spec, with eta fixed to two and the PRF call
+ * included. Creates binominally distributed elements by sampling 2*|eta| bits,
+ * and setting the coefficient to the count of the first bits minus the count of
+ * the second bits, resulting in a centered binomial distribution. Since eta is
+ * two this gives -2/2 with a probability of 1/16, -1/1 with probability 1/4,
+ * and 0 with probability 3/8.
+ */
+static
+void scalar_centered_binomial_distribution_eta_2_with_prf(scalar *out,
+                                                          const uint8_t input[33],
+                                                          ossl_mlkem_ctx *mlkem_ctx)
+{
+    uint8_t entropy[128];
+    int i;
+    uint8_t byte;
+    uint16_t value;
+
+    assert(sizeof(entropy) == 2 * /* kEta= */ 2 * DEGREE / 8);
+    prf(entropy, sizeof(entropy), input, mlkem_ctx);
```

Comment:
prf return must be checked as it can fail - IMO there is no side channel from such check